Editorial for Robot path (#3808)
cpinitiative/usaco-guide: solutions/platinum/cses-1742.mdx

---
id: cses-1742
source: CSES
title: Robot Path
author: Mihnea Brebenel
---

> ## Explanation

Reviewer: This explanation needs to be more clear. Explicitely state that the check function of the binary search is to determine if the number of intersections is equal to `[number of segments] - 1` (or whatever you're doing in your code). 

You need to specify that this is being done with a segment tree. It may be worth linking the editorial to [this problem](https://cses.fi/problemset/task/1740). 

There are also a couple implementation tricks that are worth mentioning. Firstly, any adjacent horizontal-horizontal or vertical-vertical actions can be merged, or is an upper bound on the answer. You should also mention coordinate compression and/or using an ordered set instead. You need to be clear on how your handling a) adjacent segments that intersect at their end points, and b) intersections between segments of the same direction (since `CSES intersection points forbids these types of intersections). Finally, you can directly binary search on the distance, instead of doing two binary searches on the segment, and then the distance.

Author: Ok. It's ok now ?

As you have already guessed this is a sweep line problem.
But the main issue is how to find the first intersection when the robot stops ? To find the first intersection, we use binary search.

The last thing to do is to find the length of the incomplete segment - the on that causes the intersection. There are several ways to do this.
The easiest way is to binary search for it.

## Implementation

> **Time Complexity: $\mathcal O(N \log N)$**

Reviewer: I believe that this should be $\mathcal{O}(n\log^2 n)$ because of the binary search.

Author: Ok I changed it. With the segment tree operation it's $\log^2 N$

<LanguageSection>

> <CPPSection>

Reviewer: Make sure your code passes before adding it to the module. On C++17, it's giving me a wrong answer for tests 8, 9, and 36.

Author: Working on that

---
id: cses-1742
source: CSES
title: Robot Path
author: Mihnea Brebenel
---

## Explanation
As you have already guessed this is a sweep line problem.

> Before we start talking about the actual solution, we need to make some observations about the problem's constraints. The coordinates can get easily big enough
> to cause us some issues. To solve this use [coordinate compresstion](/silver/sorting-custom?lang=cpp).

Reviewer 2: ```suggestion
Before we start talking about the actual solution, we need to make some observations about the problem's constraints. The coordinates can get easily big enough
to cause us some issues. To solve this use [coordinate compresstion](/silver/sorting-custom?lang=cpp).
```

this is really unclear - could you mention how it exceeds and what issues it would cause.

Author: Done.

> But the main issue is how to find the first intersection when the robot stops ? To find the first intersection, we use binary search. The check function in binary search uses a Segment Tree as
> a sweep line. The beginning/ending of a line is marked as an event. The events are stored in the Segment Tree.

Reviewer 2: ```suggestion
But the main issue is how to find the first intersection when the robot stops. To find the first intersection, we use binary search. The check function in binary search uses a Segment Tree as
a sweep line. The beginning/ending of a line is marked as an event. The events are stored in the Segment Tree.
```
could you reword this? it sounds more like bullet points than an explanation

> The main issue of the problem is how to find the first intersection when the robot stops ? To find the first intersection we use binary search. The check function in binary search uses a Segment Tree as
> a sweep line. The beginning/ending of a line is marked as an event. The events are stored and processed in the Segment Tree.

Reviewer 2: ```suggestion
```

Author: I reword it

> The last thing to do is to find the length of the incomplete segment - the on that causes the intersection. There are several ways to do this.
> The easiest way is to binary search its length. If you are not familiar with finding segment intersection, it may be worth solving [this](https://cses.fi/problemset/task/1740) first.

Reviewer 2: ```suggestion
The last thing to do is finding the length of the incomplete segment that causes the intersection. There are several ways to do this.

The easiest way is to binary search its length. If you are not familiar with finding segment intersections, it may be worth solving [this](https://cses.fi/problemset/task/1740) first.
```

mention the other ways too

Reviewer: You should also link the USACO Guide editorial for that problem.

Author: Done.

## Implementation

**Time Complexity: $\mathcal O(N \log^2 N)$**

<LanguageSection>
<CPPSection>
```cpp
#include <bits/stdc++.h>
#define ll long long
#define all(x) (x).begin(), (x).end()

using namespace std;

const int BEGINNING_VALUE = -2;
const int ENDING_VALUE = INT_MAX;

// BeginCodeSnip{Segment Tree Code}

> class segtree {

Reviewer 2: capitalize this class

Author: Done.

  private:
	vector<int> t;

  public:
	segtree(int N) { t.resize(3 * N); }
	void update(int node, int l, int r, int idx, int val) {
		if (l == r) {
			t[node] += val;
			return;
		}
		int mid = (l + r) >> 1;
		if (idx <= mid) update(2 * node, l, mid, idx, val);

> 		else update(2 * node + 1, mid + 1, r, idx, val);

Reviewer 2: braces

Author: Ok

		t[node] = t[2 * node] + t[2 * node + 1];
	}
	int query(int node, int l, int r, int x, int y) {
		if (r < x || l > y) return 0;
		if (x <= l && r <= y) return t[node];
		int mid = (l + r) >> 1;
		return query(2 * node, l, mid, x, y) +
		       query(2 * node + 1, mid + 1, r, x, y);
	}
};
// EndCodeSnip

struct point {
	ll x, y;
};
struct segment {
	ll a, b, x, y;

> };

Reviewer 2: could you also capitalize the structs

Author: Sure.

int n;
ll max_y;
vector<char> direction;
vector<segment> segments, original_segments;
map<char, pair<int, int>> dir;

struct event {
	ll x, y, id;
	bool operator<(const event &other) {
		if (x == other.x)
			return (0 <= id && id < n && 0 <= other.id && other.id < n)
			           ? segments[id].b < segments[other.id].b
			           : id < other.id;

		return x < other.x;
	}
};

void coordinate_compression(vector<segment> &v) {
	vector<ll> compressx, compressy;
	compressx.push_back(0);
	compressy.push_back(0);
	for (auto seg : v) {
		compressx.push_back(seg.a);
		compressx.push_back(seg.x);
		compressy.push_back(seg.b);
		compressy.push_back(seg.y);
	}
	sort(all(compressx));
	compressx.erase(unique(all(compressx)), compressx.end());
	sort(all(compressy));
	compressy.erase(unique(all(compressy)), compressy.end());
	for (segment &seg : v) {
		seg.a = lower_bound(all(compressx), seg.a) - compressx.begin() + 1;

> 		seg.x = lower_bound(all(compressx), seg.x) - compressx.begin() + 1;

Reviewer 2: space out these lines a bit

Author: Ok

		seg.b = lower_bound(all(compressy), seg.b) - compressy.begin() + 1;
		seg.y = lower_bound(all(compressy), seg.y) - compressy.begin() + 1;

		if (seg.a > seg.x) swap(seg.a, seg.x);
		if (seg.b > seg.y) swap(seg.b, seg.y);
		max_y = max(max_y, seg.b);
		max_y = max(max_y, seg.y);
	}
}

bool intersect(int idx) {
	vector<event> events;
	for (int i = 0; i <= idx; i++) {
		if (segments[i].a == segments[i].x) {
			events.push_back({segments[i].a, segments[i].b, i});
		} else {
			events.push_back({segments[i].a, segments[i].b, BEGINNING_VALUE});
			events.push_back({segments[i].x, segments[i].y, ENDING_VALUE});
		}
	}
	sort(all(events));

> 	segtree sg(max_y);

Reviewer 2: better var names plss

Author: Done

	segment last_vertical_line = {0, 0, 0, 0};
	for (auto event : events) {
		if (event.id == BEGINNING_VALUE) {

> 			// Beginnig of a horizontal line

Reviewer 2: spelling

Author: Ok

			sg.update(1, 1, max_y, event.y, 1);
			// Check for horizontal-horizontal intersection
			if (sg.query(1, 1, max_y, event.y, event.y) > 1) return true;
		} else if (event.id == ENDING_VALUE) {
			// Ending of a horizontal line
			sg.update(1, 1, max_y, event.y, -1);
		} else {
			// Checking for vertical-horizontal intersection
			if (sg.query(1, 1, max_y, segments[event.id].b,
			             segments[event.id].y) > 0)
				return true;
			// Checking for vertical-vertical intersection
			if (segments[event.id].a == last_vertical_line.a &&
			    segments[event.id].b <= last_vertical_line.y &&
			    segments[event.id].y > last_vertical_line.b)
				return true;
			last_vertical_line = segments[event.id];
		}
	}
	return false;
}

int main() {

> 	scanf("%d ", &n);

Reviewer 2: stick to cin/cout for input/output just for consistency

Author: Done.

	dir['R'] = {1, 0};
	dir['L'] = {-1, 0};
	dir['U'] = {0, 1};

> 	dir['D'] = {0, -1};

Reviewer 2: initialize with this and make it a constant

Author: ?

	ll a = 0, b = 0;
	for (int i = 0; i < n; i++) {
		char c;
		int steps;
		scanf("%c %d ", &c, &steps);
		ll x = a + dir[c].first * steps;
		ll y = b + dir[c].second * steps;
		if (i > 0) {
			a += dir[c].first;
			b += dir[c].second;
		}
		direction.push_back(c);
		segments.push_back({a, b, x, y});
		original_segments.push_back({a, b, x, y});

> 		a = x, b = y;

Reviewer 2: split into separate lines

Author: Done

	}

	coordinate_compression(segments);

	// Binary search
	int l = 0, r = n - 1, last = r;
	while (l <= r) {
		// Check if the first mid segments intersect

> 		int mid = (l + r) >> 1;

Reviewer 2: can you just do `/ 2` for everything? makes it easier to read

Author: Ok

> 		if (!intersect(mid)) l = mid + 1, last = mid;

Reviewer 2: braces

		else r = mid - 1;
	}

	auto dist = [](segment seg) {
		return abs(seg.x - seg.a) + abs(seg.y - seg.b) + 1;
	};

> 	// Check if segment a and b intersect

Reviewer 2: ```suggestion

	// Check if segment a and b intersect
```

> 	auto intersect = [&](segment a, segment b) {

Reviewer 2: ```suggestion
	auto intersect = [](segment a, segment b) {
```

Author: Done

		if (a.a > a.x) swap(a.a, a.x);
		if (a.b > a.y) swap(a.b, a.y);
		if (b.b > b.y) swap(b.b, b.y);
		if (b.a > b.x) swap(b.a, b.x);
		if (a.b == a.y && b.b == b.y && a.b == b.b)
			return a.a <= b.x && a.x > b.a;
		if (a.a == a.x && b.a == b.x && a.a == b.a)
			return a.b <= b.y && a.y > b.b;
		if (a.a == a.x && b.b == b.y)
			return a.b <= b.b && b.b <= a.y && b.a <= a.a && b.x >= a.a;
		if (a.b == a.y && b.a == b.x)
			return b.b <= a.b && a.b <= b.y && a.a <= b.a && a.x >= b.a;
		return false;
	};

	// Add up all the segment lengths before the intersection
	ll ans = -1;
	for (int i = 0; i <= last; i++) ans += dist(original_segments[i]);

	// Check if there are no intersections & if the next segment is useless
	if (last == n - 1 ||
	    (direction[last + 1] == 'D' && direction[last] == 'U') ||
	    (direction[last + 1] == 'U' && direction[last] == 'D') ||
	    (direction[last + 1] == 'L' && direction[last] == 'R') ||
	    (direction[last + 1] == 'R' && direction[last] == 'L')) {
		printf("%lli\n", ans);
		return 0;
	}

	// Find the length of the incomplete segment - the one that causes the

> 	// intersection

Reviewer 2: multiline + even lines out

	char c = direction[last + 1];
	segment seg = original_segments[last + 1];
	l = 0, r = dist(seg);
	ll add = 0;
	// Binary search its length
	while (l <= r) {
		int mid = (l + r) >> 1;
		segment new_seg = {seg.a, seg.b, seg.a + dir[c].first * mid,
		                   seg.b + dir[c].second * mid};
		bool ok = 1;
		// Check if this lenth causes an intersection with other segment
		for (int i = 0; i < last && ok; i++)
			if (intersect(new_seg, original_segments[i])) ok = 0;
		if (ok)
			l = mid + 1, add = dist(new_seg) + !(new_seg.a == new_seg.x &&
			                                     new_seg.x == new_seg.y);
		else r = mid - 1;

> 	}

Reviewer 2: braces + spelling

Author: Alright

	printf("%lli\n", ans + add);

> 
> 	return 0;

Reviewer 2: ```suggestion
```

---
id: cses-1742
source: CSES
title: Robot Path
author: Mihnea Brebenel
---

> ## Solution 1 - Using Segment Tree

Reviewer 3: I think your solution doesn't completely explain to an outside reader how to solve the problem, so I think you should rewrite this also using a more formal tone and better grammar (Use Grammarly).

> As you have already guessed this is a sweep line problem. In this approach I handle the sweep line with a Segment Tree.

Reviewer 3: No need for this first line

Author: Ok removed

> Before we start talking about the actual solution, we need to make some observations about the problem's constraints. The coordinates can get easily big enough

Reviewer 3: ```suggestion
First, we need to make some observations about the problem's constraints. The coordinates can get easily big enough
```

> Before we start talking about the actual solution, we need to make some observations about the problem's constraints. The coordinates can get easily big enough
> to cause us some issues like - the coordinates values can reach $10^11$ way too big to be handled by Segment Tree. To solve this use [coordinate compresstion](/silver/sorting-custom?lang=cpp).

Reviewer 3: ```suggestion
First, we need to make some observations about the problem's constraints. The coordinates values can reach $10^11$, which is beyond the memory limit of a segment tree. To solve this, use [coordinate compresstion](/silver/sorting-custom?lang=cpp).
```

Author: Ok changed

> Now with the coordinates compressed we can easily use the sweep line. The goal is to find the *first* intersection in the robot path. To find the first intersection we use binary search. The check function in binary search takes the first mid segments

Reviewer 3: > Now with the coordinates compressed we can easily use the sweep line.

Reword this sentence

Reviewer 3: > The goal is to find the *first* intersection in the robot path. To find the first intersection we use binary search. The check function in binary search takes the first mid segments

Combine these three sentences into one shorter sentence

> and determines if there is an intersection .The beginning/ending of a line is marked as an event which are stored and processed by the Segment Tree.

Reviewer 3: ```suggestion
and determines if there is an intersection. The beginning and end of a line are marked as an event, which is stored and processed by the segment tree.
```

Author: Ok

> The last thing to do is to find the length of the incomplete segment - the on that causes the intersection. There are several ways to do this.

Reviewer 3: This doesnt make sense - reword it

One way is to use again the sweep line to find the first intersection of the line, but the implementation would be tricky. You could also use brute force and try for every possible length as long as it doesn't intersect another segment.

> We can improve the brute force approach by using binary search. If you are not familiar with finding segment intersection, it may be worth solving [this](https://cses.fi/problemset/task/1740) first.

Reviewer 3: combine these three ideas into a sentence or two

Reviewer 3: also you didn't explain the binary search approach...

## Solution 2 - Using set

> In this approach the sweep line is handled by a set instead of a segment tree, and thus the coordinate compression is no needed anymore.

Reviewer 3: ```suggestion
In this approach, the sweep line is handled by a set instead of a segment tree; thus, coordinate compression is no longer needed.
```

## Implementation

**Time Complexity: $\mathcal O(N \log^2 N)$**

> <LanguageSection>

Reviewer 3: I recommend adding more comments to your code - part of it is hard to understand

using namespace std;

const int BEGINNING_VALUE = -2;
const int ENDING_VALUE = INT_MAX;

// BeginCodeSnip{Segment Tree Code}
class Segtree {
  private:
	vector<int> t;

  public:
	Segtree(int N) { t.resize(3 * N); }
	void update(int node, int l, int r, int idx, int val) {
		if (l == r) {
			t[node] += val;
			return;
		}
		int mid = (l + r) / 2;
		if (idx <= mid) {
			update(2 * node, l, mid, idx, val);
		} else {
			update(2 * node + 1, mid + 1, r, idx, val);
		}
		t[node] = t[2 * node] + t[2 * node + 1];
	}
	int query(int node, int l, int r, int x, int y) {
		if (r < x || l > y) return 0;
		if (x <= l && r <= y) return t[node];
		int mid = (l + r) / 2;
		return query(2 * node, l, mid, x, y) +
		       query(2 * node + 1, mid + 1, r, x, y);
	}
};
// EndCodeSnip

struct Point {
	ll x, y;
};
struct Segment {
	ll a, b, x, y;
};

int n;
ll max_height;
vector<char> direction;
vector<Segment> segments, original_segments;
map<char, pair<int, int>> dir;

struct event {
	ll x, y, id;
	bool operator<(const event &other) {
		if (x == other.x)
			return (0 <= id && id < n && 0 <= other.id && other.id < n)
			           ? segments[id].b < segments[other.id].b
			           : id < other.id;

> 		return x < other.x;

Reviewer 3: put this in an if else statement

	compressx.push_back(0);
	compressy.push_back(0);

	for (auto seg : v) {
		compressx.push_back(seg.a);
		compressx.push_back(seg.x);
		compressy.push_back(seg.b);
		compressy.push_back(seg.y);
	}

	sort(all(compressx));
	compressx.erase(unique(all(compressx)), compressx.end());
	sort(all(compressy));
	compressy.erase(unique(all(compressy)), compressy.end());

	for (Segment &seg : v) {
		seg.a = lower_bound(all(compressx), seg.a) - compressx.begin() + 1;
		seg.x = lower_bound(all(compressx), seg.x) - compressx.begin() + 1;

		seg.b = lower_bound(all(compressy), seg.b) - compressy.begin() + 1;
		seg.y = lower_bound(all(compressy), seg.y) - compressy.begin() + 1;

		if (seg.a > seg.x) swap(seg.a, seg.x);
		if (seg.b > seg.y) swap(seg.b, seg.y);
		max_height = max(max_height, seg.b);
		max_height = max(max_height, seg.y);
	}
}

bool intersect(int idx) {
	vector<event> events;
	for (int i = 0; i <= idx; i++) {
		if (segments[i].a == segments[i].x) {
			events.push_back({segments[i].a, segments[i].b, i});
		} else {
			events.push_back({segments[i].a, segments[i].b, BEGINNING_VALUE});
			events.push_back({segments[i].x, segments[i].y, ENDING_VALUE});
		}
	}
	sort(all(events));

> 

Reviewer 3: inconsistent with the spacing throughout the rest of the method

	Segtree sg(max_height);
	Segment last_vertical_line = {0, 0, 0, 0};
	for (auto event : events) {
		if (event.id == BEGINNING_VALUE) {
			// Beginning of a horizontal line
			sg.update(1, 1, max_height, event.y, 1);
			// Check for horizontal-horizontal intersection
			if (sg.query(1, 1, max_height, event.y, event.y) > 1) return true;
		} else if (event.id == ENDING_VALUE) {
			// Ending of a horizontal line
			sg.update(1, 1, max_height, event.y, -1);
		} else {
			// Checking for vertical-horizontal intersection
			if (sg.query(1, 1, max_height, segments[event.id].b,
			             segments[event.id].y) > 0)

> 				return true;

Reviewer 3: This looks weird

			// Checking for vertical-vertical intersection
			if (segments[event.id].a == last_vertical_line.a &&
			    segments[event.id].b <= last_vertical_line.y &&
			    segments[event.id].y > last_vertical_line.b)

> 				return true;

Reviewer 3: This too

		ll x = a + dir[c].first * steps;
		ll y = b + dir[c].second * steps;

		if (i > 0) {
			a += dir[c].first;
			b += dir[c].second;
		}

		direction.push_back(c);

		segments.push_back({a, b, x, y});
		original_segments.push_back({a, b, x, y});

		a = x, b = y;
	}

	coordinate_compression(segments);

	// Binary search
	int l = 0, r = n - 1, last = r;
	while (l <= r) {
		// Check if the first mid segments intersect
		int mid = (l + r) >> 1;
		if (!intersect(mid)) l = mid + 1, last = mid;
		else r = mid - 1;
	}

	auto dist = [](Segment seg) {
		return abs(seg.x - seg.a) + abs(seg.y - seg.b) + 1;
	};
	// Check if segment a and b intersect
	auto intersect = [](Segment a, Segment b) {
		if (a.a > a.x) swap(a.a, a.x);
		if (a.b > a.y) swap(a.b, a.y);
		if (b.b > b.y) swap(b.b, b.y);
		if (b.a > b.x) swap(b.a, b.x);
		if (a.b == a.y && b.b == b.y && a.b == b.b)
			return a.a <= b.x && a.x > b.a;
		if (a.a == a.x && b.a == b.x && a.a == b.a)
			return a.b <= b.y && a.y > b.b;
		if (a.a == a.x && b.b == b.y)
			return a.b <= b.b && b.b <= a.y && b.a <= a.a && b.x >= a.a;
		if (a.b == a.y && b.a == b.x)
			return b.b <= a.b && a.b <= b.y && a.a <= b.a && a.x >= b.a;
		return false;

> 	};

Reviewer 3: Is there a better way to check if they intersect without writing 15 lines of code?
